### PR TITLE
Reworked redriver. 

### PR DESCRIPTION
**Pull request by gauravAshok**
Earlier, if the node holding the singleton of MessageScheduler were to go down, the cluster would stop processing any request, as the register/deregister calls would stuck.

* Moved register/deregister methods out of the messageScheduler. Now every node serves its own register/deregister requests.
* RedriverRegisterImpl takes care of the register/deregister calls by directing it to MessageManagerService.
* Renamed messageScheduler to RedriverService. It's job now only is to periodically check DB for oldest messages and redrive it if there is a need.

### CODE REVIEW

File: redriver/src/main/java/com/flipkart/flux/redriver/service/MessageManagerService.java [hunk 1]
```diff
@@ -39,70 +40,74 @@
...
     public MessageManagerService(MessageDao messageDao,
+                                 @Named("redriver.noOfPersistenceWorkers") int noOfPersistenceWorkers,
                                  @Named("redriver.batchdelete.intervalms") Integer batchDeleteInterval,
                                  @Named("redriver.batchdelete.batchsize") Integer batchSize) {
         this.messageDao = messageDao;
-        this.batchDeleteInterval = Long.valueOf(batchDeleteInterval);
+        this.batchDeleteInterval = batchDeleteInterval;
         this.batchSize = batchSize;
         this.messagesToDelete = new ConcurrentLinkedQueue<>();
         scheduledExecutorService =
```
Comment (shyamakirala) [inline redriver/src/main/java/com/flipkart/flux/redriver/service/MessageManagerService.java]: scheduledDeletionService ?

File: redriver/src/main/java/com/flipkart/flux/redriver/service/MessageManagerService.java [hunk 2]
```diff
@@ -39,70 +40,74 @@
...
-            new InstrumentedScheduledExecutorService(Executors.newScheduledThreadPool(2), SharedMetricRegistries.getOrCreate(METRIC_REGISTRY_NAME));
+            new InstrumentedScheduledExecutorService(Executors.newScheduledThreadPool(2), SharedMetricRegistries.getOrCreate(METRIC_REGISTRY_NAME), scheduledExectorSvcName);
+        persistenceExecutorService = Executors.newFixedThreadPool(noOfPersistenceWorkers);
```
Comment (shyamakirala) [inline redriver/src/main/java/com/flipkart/flux/redriver/service/MessageManagerService.java]: I think we should change it to InstrumentedExecutorService to get metrics

File: redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static com.flipkart.flux.Constants.METRIC_REGISTRY_NAME;
+
+/**
+ * The service uses a scheduler to read the oldest message with fixed delay and redrives them if necessary. The task will
+ * be scheduled for execution if the current time is greater than or equal to scheduledTime.
```
Comment (shyamakirala) [inline redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java]: @author :)

File: runtime/src/main/resources/packaged/configuration.yml [hunk 4]
```diff
@@ -48,7 +48,9 @@ redriver:
   batchdelete:
     batchsize: 50
     intervalms: 1000
-  noOfRedriverWorkers : 50
+  batchread:
+    batchsize: 100
+    intervalms: 20000
```
Comment (shyamakirala) [inline runtime/src/main/resources/packaged/configuration.yml]: we may need to reduce it. Will do a perf before putting in prod, and fix this.

File: redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        int offset = 0;
+        List<ScheduledMessage> messages;
+        long now = System.currentTimeMillis();
+        do {
+            messages = messageService.retrieveOldest(offset, batchSize);
+            messages.stream().filter(e -> e.getScheduledTime() < now).forEach(e -> redriverRegistry.redriveTask(e.getTaskId()));
+            offset += batchSize;
+            // get next batch if we found batchSize tasks and all were redriven
+        } while (messages.size() == batchSize && messages.get(messages.size() - 1).getScheduledTime() < now);
+    }
```
Comment (gauravAshok) [inline redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java]: you mean, if there are only few total msgs in db, then we clear them up regardless of their timings. OR if we found few msgs that can be redriven then we should do that. the latter is already happening.

File: redriver/src/main/java/com/flipkart/flux/redriver/service/MessageManagerService.java (hunk 1 above)
Comment (gauravAshok) [inline redriver/src/main/java/com/flipkart/flux/redriver/service/MessageManagerService.java]: are you suggesting the class name?

File: redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java (hunk 5 above)
Comment (shyamakirala) [inline redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java]: was talking about the second case. My bad, didn't see it.

**Comment by shyamakirala** [comment]
Looks good.

**Pull request merged**